tools/kagami/src/crypto.rs

```diff
@@ -1,14 +1,135 @@
 
+#[derive(Subcommand, Debug, Clone)]
+pub enum CryptoMode {
+    GenesisSigning(GenesisSigningArgs),
+    KeyPairGeneration(KeyPairArgs),
+}
+
+/// Use `Kagami` to sign genesis block.
+#[derive(ClapArgs, Clone, Debug)]
+#[command(group = ArgGroup::new("private_key").required(true))]
+#[command(group = ArgGroup::new("public_key").required(true))]
+#[command(group = ArgGroup::new("format").required(false))]
+pub struct GenesisSigningArgs {
+    /// The algorithm of the provided keypair
+    #[clap(default_value_t, long, short)]
+    algorithm: AlgorithmArg,
+    /// Private key (in string format) to sign genesis block
+    #[clap(long, group = "private_key")]
+    private_key_string: Option<String>,
+    /// Path to private key to to sign genesis block
+    #[clap(long, group = "private_key")]
+    private_key_path: Option<PathBuf>,
+    /// Public key of the corresponding private key
+    #[clap(long, group = "public_key")]
+    public_key_string: Option<String>,
+    /// Path to public key of the corresponding private key
+    #[clap(long, group = "public_key")]
+    public_key_path: Option<PathBuf>,
+    /// Path to json-serialized keypair
+    #[clap(long, short, group = "private_key", group = "public_key")]
+    keypair_path: Option<PathBuf>,
+    /// Unique id of blockchain
+    #[clap(long)]
+    chain_id: ChainId,
+    /// Path to genesis json file
+    #[clap(long, short)]
+    genesis_path: PathBuf,
+    /// Path to signed genesis output file
+    #[clap(long, short, group = "format")]
+    output_path: Option<PathBuf>,
+}
+
+impl GenesisSigningArgs {
+    fn get_private_key(&self) -> Result<PrivateKey> {
+        let private_key_bytes =
+            Self::get_key_bytes(&self.private_key_path, &self.private_key_string)?;
+        PrivateKey::from_bytes(self.algorithm.0, private_key_bytes.borrow()).wrap_err_with(|| {
+            eyre!(
+                "Failed to parse private key for algorithm `{}`",
+                self.algorithm
+            )
+        })
+    }
+
+    fn get_public_key(&self) -> Result<PublicKey> {
+        let public_key_bytes = Self::get_key_bytes(&self.public_key_path, &self.public_key_string)?;
+        PublicKey::from_bytes(self.algorithm.0, public_key_bytes.borrow()).wrap_err_with(|| {
+            eyre!(
+                "Failed to parse public key for algorithm `{}`",
+                self.algorithm
+            )
+        })
+    }
+
+    fn get_key_bytes<'a, P: AsRef<Path>>(
+        path: &Option<P>,
+        value: &'a Option<String>,
+    ) -> Result<Cow<'a, [u8]>, std::io::Error> {
+        match (path, value) {
+            (Some(path_buf), None) => Ok(Cow::Owned(fs::read(path_buf)?)),
+            (None, Some(hex)) => Ok(Cow::Borrowed(hex.as_bytes())),
+            _ => unreachable!("Clap group invariant"),
+        }
+    }
+
+    fn read_keypair<P: AsRef<Path>>(path: P) -> Result<KeyPair> {
+        let bytes = fs::read(path)?;
+        Ok(serde_json::from_slice(bytes.as_slice())?)
+    }
```

Comment:
Minor thing: these 2 methods don't seem to be related to `GenesisSigningArgs` struct. Consider externalising as plain `fn`s.